File: 백준 N과 M (1) -백트랙킹-/복습/Main.java
import java.util.Scanner;

public class Main {
    static int N;
    static int M;
    static boolean[] isVisited;
    static int[] num;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        isVisited = new boolean[N + 1];
        num = new int[N + 1];
        dfs(1);

    }

    static void dfs(int current) {
        if (current > M) {
            for (int i = 1; i <= M; i++) {
                System.out.print(num[i] + " ");
            }
            System.out.println();
            return;
        }
        for (int i = 1; i <= N; i++) {
            if (isVisited[i]) {
                continue;
            }
            isVisited[i] = true;
            num[current] = i;
            dfs(current + 1);
            isVisited[i] = false;
        }
    }


}

File: 백준 N과 M (1) -백트랙킹-/복습/Main.kt
import java.util.*

internal var N: Int = 0
internal var M: Int = 0
internal var isVisited: BooleanArray? = null
internal var num: IntArray? = null


fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    N = sc.nextInt()
    M = sc.nextInt()
    isVisited = BooleanArray(N + 1)
    num = IntArray(N + 1)
    dfs(1)

}

internal fun dfs(current: Int) {
    if (current > M) {
        for (i in 1..M) {
            print(num!![i].toString() + " ")
        }
        println()
        return
    }
    for (i in 1..N) {
        if (isVisited!![i]) {
            continue
        }
        isVisited!![i] = true
        num!![current] = i
        dfs(current + 1)
        isVisited!![i] = false
    }
}


